src/domain/spaces/address-books/address-book-items.repository.ts

```diff
@@ -1,54 +1,95 @@
+import { AddressBookItem } from '@/domain/spaces/address-books/entities/address-book-item.entity';
 import { Space } from '@/domain/spaces/entities/space.entity';
 import { ISpacesRepository } from '@/domain/spaces/spaces.repository.interface';
 import { IUsersRepository } from '@/domain/users/users.repository.interface';
-import { Inject, Injectable, UnauthorizedException } from '@nestjs/common';
-import { In } from 'typeorm';
+import {
+  BadRequestException,
+  Inject,
+  Injectable,
+  UnauthorizedException,
+} from '@nestjs/common';
+import { EntityManager, In } from 'typeorm';
+import { UpsertAddressBookItemsDto } from '@/routes/spaces/entities/upsert-address-book-items.dto.entity';
+import { MemberRole } from '@/domain/users/entities/member.entity';
 
 @Injectable()
 export class AddressBookItemsRepository implements IAddressBookItemsRepository {
+  private readonly maxItems: number;
+
   constructor(
     private readonly db: PostgresDatabaseService,
     @Inject(ISpacesRepository)
     private readonly spacesRepository: ISpacesRepository,
     @Inject(IUsersRepository)
     private readonly userRepository: IUsersRepository,
-  ) {}
+    @Inject(IConfigurationService)
+    private readonly configurationService: IConfigurationService,
+  ) {
+    this.maxItems = this.configurationService.getOrThrow<number>(
+      'spaces.addressBooks.maxItems',
+    );
+  }
 
-  async findAllBySpaceId(args: {
+  public async findAllBySpaceId(args: {
     authPayload: AuthPayload;
     spaceId: Space['id'];
-  }): Promise<Array<AddressBookItem>> {
-    const space = await this.getSpaceAsMember(args);
+  }): Promise<Array<AddressBookDbItem>> {
+    const space = await this.getSpaceAs({
+      ...args,
+      memberRoleIn: ['ADMIN', 'MEMBER'],
+    });
     const repository = await this.db.getRepository(DbAddressBookItem);
     return repository.findBy({ space: { id: space.id } });
   }
 
-  // eslint-disable-next-line @typescript-eslint/no-unused-vars
-  upsertMany(args: {
+  public async upsertMany(args: {
     authPayload: AuthPayload;
     spaceId: Space['id'];
-    addressBookItems: Array<AddressBookItem>;
-  }): Promise<Array<AddressBookItem>> {
-    throw new Error('Method not implemented.');
+    addressBookItems: UpsertAddressBookItemsDto['items'];
+  }): Promise<Array<AddressBookDbItem>> {
+    const space = await this.getSpaceAs({
+      ...args,
+      memberRoleIn: ['ADMIN'],
+    });
+    const repository = await this.db.getRepository(DbAddressBookItem);
+    await this.db.transaction(async (entityManager) => {
+      const existingAddresses = await this.updateExistingAddressBookItems({
+        entityManager,
+        addressBookItems: args.addressBookItems,
+        space,
+        authPayload: args.authPayload,
+      });
+      await this.createNewAddressBookItems({
+        entityManager,
+        addressBookItems: args.addressBookItems.filter(
+          (item) => !existingAddresses.includes(item.address),
```

Comment:
Nit: let's use `isAddressEqual` here to account for a mixture of (non-)checksummed addresses (although it shouldn't happen).